accepted/single-file/design.md

```diff
@@ -66,225 +32,254 @@ Publishing to a single file can be triggered by adding the following property to
 
-### New API
+There are a few options to consider for the `Assembly.Location` property of a bundled assembly:
 
-In this section, we propose adding a few APIs to facilitate common operations on bundled-apps.
+* A fixed literal (ex: `null`) indicating that no actual location is available.
+* The simple name of the assembly (with no path).
+* The path of the assembly as if it were not to be packaged into the single-file.
+* A special UNC notation such as `<bundle-path>/:/asm.dll` to denote files that come from the bundle. 
+* A configurable selection of the above, etc.
 
-The binaries that are published in the project are expected to be handled transparently by the host. However, explicit access to the embedded files is useful in situations such as:
+Proposed solution is: `:/<bundle-sub-path>/asm.dll`, where the `:` denotes the location within the bundle. The `:` is not a valid character in file-names on Windows. The `:` is already used as a path-separator on Unix systems.
 
-- Reading additional files packaged into the app (ex: data files).
-- Open an assembly for reflection/inspection
-- Load plugins built as single-file class-libs using existing Loader APIs.
+Most of the app development can be agnostic to whether the app is published as single-file or not. However, the parts of the app that deal with physical locations of files need to be aware of the single-file packaging. 
+
+#### `AppContext.BaseDirectory`
 
-Therefore, we propose adding an API similar to [GetManifestResourceStream](https://docs.microsoft.com/en-us/dotnet/api/system.reflection.assembly.getmanifestresourcestream?view=netframework-4.7.2#System_Reflection_Assembly_GetManifestResourceStream_System_String_) to obtain a stream corresponding to an embedded file. 
+`AppContext.BaseDirectory` will be the directory where the AppHost (the single-file bundle itself) resides.
 
-This is only a draft of the proposed APIs. The actual shape of the APIs will be decided via API review process.
+## Handling Content 
+
+The app may want to access certain embedded content for reading, rather than loading via host/runtime. For example: bundled payload/data files. In this case, the recommended strategy is to embed the content files within appropriate managed assemblies as PE resources, and access them through [resource handling APIs](https://docs.microsoft.com/en-us/dotnet/api/system.reflection.assembly.getmanifestresourceinfo?view=netcore-3.1). 
+
+However, explicit access to the embedded files is useful in situations such as: 
+- Reading data-files from the app's native code
+- Open an assembly for reflection/inspection
+
+Therefore, we propose adding an API similar to [GetManifestResourceStream](https://docs.microsoft.com/en-us/dotnet/api/system.reflection.assembly.getmanifestresourcestream?view=netcore-3.1) to obtain a stream corresponding to an embedded file. This is only a draft of the proposed APIs. The actual shape of the APIs will be decided via API review process.
 
```

Comment:
The one API that I think would make sense to add is an API that returns launching .exe path. It is clearly defined what it does and you can do that today using System.Diagnostics.Process, but it is show and comes with a lot of unnecessary dependencies. Some discussion on this is in https://github.com/dotnet/runtime/issues/3704 .

Reply:
For single-exe in .net 5, the AppContext.BaseDirectory will point to the launching EXE path.
The launching EXE will be the app.exe within this directory.

However, there are other scenarios for launching the app (ex: dotnet run) where the launching EXE is not in the same directory as the app. So, is your recommendation to find the launching EXE in these cases?

Reply:
> AppContext.BaseDirectory will point to the launching EXE path

It sounds good to me.

> So, is your recommendation to find the launching EXE in these cases?

It depends what you need the launching .exe for. For example, if you would like to relaunch the app under different credentials or with extra command line arguments, you do want to use the path to dotnet.exe in your example.